File: camerax_android/android/src/main/kotlin/dev/hebei/camerax_android/common/ConsumerImpl.kt
package dev.hebei.camerax_android.common

import androidx.camera.mlkit.vision.MlKitAnalyzer
import androidx.camera.video.VideoRecordEvent
import dev.hebei.camerax_android.CameraXImpl
import dev.hebei.camerax_android.PigeonApiMlKitAnalyzerResultConsumerApi
import dev.hebei.camerax_android.PigeonApiVideoRecordEventConsumerApi

class VideoRecordEventConsumerImpl(impl: CameraXImpl) : PigeonApiVideoRecordEventConsumerApi(impl) {
    override fun pigeon_defaultConstructor(): VideoRecordEventConsumer {
        return VideoRecordEventConsumer(this)
    }
}

class MlKitAnalyzerResultConsumerImpl(impl: CameraXImpl) : PigeonApiMlKitAnalyzerResultConsumerApi(impl) {
    override fun pigeon_defaultConstructor(): MlKitAnalyzerResultConsumer {
        return MlKitAnalyzerResultConsumer(this)
    }
}

class VideoRecordEventConsumer(private val impl: VideoRecordEventConsumerImpl) :
    androidx.core.util.Consumer<VideoRecordEvent> {
    override fun accept(value: VideoRecordEvent) {
        impl.accept(this, value) {}
    }
}

class MlKitAnalyzerResultConsumer(private val impl: MlKitAnalyzerResultConsumerImpl) :
    androidx.core.util.Consumer<MlKitAnalyzer.Result> {
    override fun accept(value: MlKitAnalyzer.Result) {
        impl.accept(this, value) {}
    }
}

File: camerax_android/lib/src/common/consumer_impl.dart
import 'package:camerax_android/src/camerax_api.g.dart';
import 'package:camerax_android/src/core.dart';
import 'package:camerax_android/src/ml.dart';
import 'package:camerax_android/src/video.dart';
import 'package:camerax_platform_interface/camerax_platform_interface.dart';

abstract base class ConsumerImpl<T> implements Consumer<T> {}

final class ImageProxyConsumerImpl extends ConsumerImpl<ImageProxy> {
  ImageProxyConsumerProxyApi api;

  ImageProxyConsumerImpl.internal(this.api);
}

final class MlKitAnalyzerResultConsumerImpl
    extends ConsumerImpl<MlKitAnalyzer$Result> {
  final MlKitAnalyzerResultConsumerProxyApi api;

  MlKitAnalyzerResultConsumerImpl.internal(this.api);
}

final class VideoRecordEventConsumerImpl
    extends ConsumerImpl<VideoRecordEvent> {
  final VideoRecordEventConsumerProxyApi api;

  VideoRecordEventConsumerImpl.internal(this.api);
}

final class OtherConsumerImpl<T> extends ConsumerImpl<T> {
  final void Function(T value) accept;

  OtherConsumerImpl({required this.accept});
}

final class ConsumerChannelImpl extends ConsumerChannel {
  @override
  Consumer<T> create<T>({required void Function(T value) accept}) {
    if (T == ImageProxy) {
      final api = ImageProxyConsumerProxyApi(
        accept: (_, e) => accept(e.impl as T),
      );
      return ImageProxyConsumerImpl.internal(api) as ConsumerImpl<T>;
    }
    if (T == MlKitAnalyzer$Result) {
      final api = MlKitAnalyzerResultConsumerProxyApi(
        accept: (_, e) => accept(e.impl as T),
      );
      return MlKitAnalyzerResultConsumerImpl.internal(api) as ConsumerImpl<T>;
    }
    if (T == VideoRecordEvent) {
      final api = VideoRecordEventConsumerProxyApi(
        accept: (_, e) => accept(e.impl as T),
      );
      return VideoRecordEventConsumerImpl.internal(api) as ConsumerImpl<T>;
    }
    return OtherConsumerImpl(accept: accept);
  }
}

extension ConsumerX<T> on Consumer<T> {
  void accept(T value) {
    final impl = this;
    if (impl is! OtherConsumerImpl<T>) throw TypeError();
    impl.accept(value);
  }
}

extension ImageProxyConsumerX on Consumer<ImageProxy> {
  ImageProxyConsumerProxyApi get api {
    final impl = this;
    if (impl is! ImageProxyConsumerImpl) throw TypeError();
    return impl.api;
  }
}

extension MlKitAnalyzer$ResultConsumerX on Consumer<MlKitAnalyzer$Result> {
  MlKitAnalyzerResultConsumerProxyApi get api {
    final impl = this;
    if (impl is! MlKitAnalyzerResultConsumerImpl) throw TypeError();
    return impl.api;
  }
}

extension VideoRecordEventConsumerX on Consumer<VideoRecordEvent> {
  VideoRecordEventConsumerProxyApi get api {
    final impl = this;
    if (impl is! VideoRecordEventConsumerImpl) throw TypeError();
    return impl.api;
  }
}
